loam/evaluation.lisp

```diff
@@ -47,166 +47,231 @@
-  (rule (cons-rel car cdr cons) <--
-    (cons-mem car cdr addr)
-    (let ((cons (ptr :cons (dual-value addr))))))
-
-  ;; signal
-  (rule (unhash4 digest) <--
-    (ingress ptr) (when (has-tag-p ptr :cons)) (ptr-value ptr digest))
-
-  ;; signal
-  (rule (hash4 car-tag car-value cdr-tag cdr-value) <--
-    (egress cons)
-    (cons-rel car cdr cons)
-    (let ((car-tag (widen (ptr-tag car)))
-	  (cdr-tag (widen (ptr-tag cdr)))))
-    (ptr-value car car-value) (ptr-value cdr cdr-value)
-    #+nil(let ((x (prog-trace 'hash4-cons car-tag car-value cdr-tag cdr-value))))
-    )
-
-  ;; signal
-  (rule (egress car) (egress cdr) <--
-    (egress cons) (cons-rel car cdr cons)))
-
-;; FIXME: Align with Lurk by dropping closed-env-tag in the hash.
-(defprogram thunk-mem (hash-cache)
-  (include ptr-program)
-  (include hash4)
-
-  ; signal
-  (relation (thunk ptr ptr)) ; (body closed-env)
-
-  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-  ;;; Memory
-
-  ;; The canonical thunk Ptr relation.
-  (relation (thunk-rel ptr ptr ptr)) ; (body closed-env thunk)
-
-  ;; Memory to support thunks allocated by digest or contents.
-  (lattice (thunk-digest-mem wide dual-element)) ; (digest addr)
-  (lattice (thunk-mem ptr ptr dual-element)) ; (body closed-env addr)
-
-  ;; Populating alloc(...) triggers allocation in thunk-digest-mem.
-  (rule (thunk-digest-mem value  (alloc :thunk (dual 0))) <--
-    (alloc (tag-address :thunk) value))
-
-  ;; Populating thunk(...) triggers allocation in thunk-mem.
-  (rule (thunk-mem body closed-env (alloc :thunk (dual 0))) <-- (thunk body closed-env))
-
-  ;; Populate thunk-digest-mem if a thunk in thunk-mem has been hashed in hash4-rel.
-  (rule (thunk-digest-mem digest addr) <--
-    (thunk-mem body closed-env addr)
-    (ptr-value body body-value) (ptr-value closed-env closed-env-value)
-    (hash4-rel (widen (ptr-tag body)) body-value (widen (ptr-tag closed-env)) closed-env-value digest))
-
-  ;; Other way around.
-  (rule (thunk-mem body closed-env addr) <--
-    (thunk-digest-mem digest addr)
-    (hash4-rel body-tag body-value closed-env-tag closed-env-value digest)
-    (ptr-value body body-value) (ptr-value closed-env closed-env-value)
-    (when (and (== (ptr-tag body) (wide-nth 0 body-tag))
-	       (== (ptr-tag closed-env) (wide-nth 0 closed-env-tag)))))
-
-  ;; Register a thunk value.
-  (rule (ptr-value thunk value) <--
-    (thunk-digest-mem value addr) (let ((thunk (ptr :thunk (dual-value addr))))))
-
-  ;; Register a thunk relation.
-  (rule (thunk-rel body closed-env thunk) <--
-    (thunk-mem body closed-env addr)
-    (let ((thunk (ptr :thunk (dual-value addr))))))
+  )
 
-  ;; signal
-  (rule (unhash4 digest) <--
-    (ingress ptr) (when (has-tag-p ptr :thunk)) (ptr-value ptr digest))
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  ;; TODO: Any better way to do this?
+  (defun concat-sym (root suf)
+    (intern (format nil "~A~A" root suf)))
```

Comment:
You can use [symbolconc](https://github.com/lurk-lab/loam/blob/main/base/macros.lisp#L19-L20)